### PR TITLE
Navneeth add implementation for resolveTask permission

### PR DESCRIPTION
**Pull request by navneeeth**
# Description
task description

## Related PRs (if any):
Please test the PR with the latest development branch on the backend.

## Main changes explained:
- Update check for whether an user is able to resolve a task.
- Call a dispatch to check if the user has the `resolveTask` permission.
- Add a toast message on successful 'mark as done' operation on a task.
- Remove unused state variable `shouldFetchData` in `src/components/Timelog/Timelog.jsx

## How to test:
1. Check into the current branch.
2. Make the following code change in `src/components/TeamMemberTasks/TeamMemberTask.jsx` for easier testing:
- Replace line 65: 
```javascript
const rolesAllowedToResolveTasks = ['Administrator', 'Owner'];
```
with: 
```javascript
const rolesAllowedToResolveTasks = ['Owner'];
```
4. Run the `development` branch of the backend with `npm install` and `npm run dev`.
5. Run this frontend branch with `npm install` and `npm run start:local`.
6. Clear site data/cache.
7. Log in as Owner -> Other Links -> Permissions Management -> Manage User Permissions.
8. Enter the username of an Admin user that does not have the “Resolve Tasks” permission. Add the permission and Click on Submit.
9. Login as the chosen Admin user (if already logged in, make sure to log out and log back in). The user should be able to see the ‘green tick’ icon for tasks displayed on Dashboard -> Tasks.
10. Resolve a task successfully by clicking on the green tick and verify if it is no longer present. See the third test video 'Adding a new task' for instructions on creating a new task.

### Code Changes
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/18de107e-a284-45b7-b96e-56947df9e4e4

### Testing
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/82faaf12-6379-44e2-8dec-d1562024586d

### Adding a new task

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/a66b5ed6-614e-41a3-8133-13cba00eb3fc

### CODE REVIEW

File: src/components/Timelog/Timelog.jsx [hunk 1]
```diff
@@ -117,7 +117,7 @@ const Timelog = props => {
     personId: displayUserProfile._id,
   }
 
-  // const [shouldFetchData, setShouldFetchData] = useState(false);
+  const [shouldFetchData, setShouldFetchData] = useState(false);
```
Comment (Abi-Liu) [inline src/components/Timelog/Timelog.jsx:120]: Remove this state as it's not being used anywhere.

File: src/components/Timelog/Timelog.jsx [hunk 2]
```diff
@@ -392,14 +392,14 @@ const Timelog = props => {
     }
   }, [timeLogState.isTimeEntriesLoading, timeEntries]);
 
-  useEffect(() => {
+   useEffect(() => {
       loadAsyncData(displayUserId);
-  }, [displayUserId]);
+   }, [displayUserId]);
 
   // useEffect(() => {
```
Comment (Abi-Liu) [inline src/components/Timelog/Timelog.jsx:399]: Remove excess code.

**Review by Abi-Liu — CHANGES_REQUESTED** [review]
Overall it works as expected. I followed the testing instructions to allow my admin account to resolve tasks and I can now see and click on the green checkmarks to resolve tasks.

There are some areas where the code can be cleaned up, unused state variables, commented out useEffects, etc.

Some more areas that can be cleaned up outside of the comments I have left are the `handeUpdateTask` function inside of `src/components/TeamMemberTasks/TeamMemberTask.jsx` and `src/components/Timelog/Timelog.jsx`

That function is just setting the `shouldFetchData` state to true but that piece of state is not being used anywhere so it's just causing unnecessary rerenders.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/59585815/d43e88fc-1350-49da-9bc9-0d176de51bd7

**Comment by navneeeth** [comment]
> Overall it works as expected. I followed the testing instructions to allow my admin account to resolve tasks and I can now see and click on the green checkmarks to resolve tasks.
> 
> There are some areas where the code can be cleaned up, unused state variables, commented out useEffects, etc.
> 
> Some more areas that can be cleaned up outside of the comments I have left are the `handeUpdateTask` function inside of `src/components/TeamMemberTasks/TeamMemberTask.jsx` and `src/components/Timelog/Timelog.jsx`
> 
> That function is just setting the `shouldFetchData` state to true but that piece of state is not being used anywhere so it's just causing unnecessary rerenders.
> 
>  video1942620258.mp4

Hello, @Abi-Liu! Thanks for your thorough review. I have removed the ununsed state variable and some additional comments. Requesting you to review the PR again. Thanks!

**Review by Abi-Liu — APPROVED** [review]
Looks good!

**Review by Pandani07 — APPROVED** [review]
PR works as expected. I added the permission to my admin account and I could see the green tick mark to resolve tasks

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/48dd2f0d-c5c3-4f51-91f2-685c917fa112

**Review by Coops023 — APPROVED** [review]
Hi Navneeth, I tested this PR and after granting resolve task permission to my admin account i was able to resolve tasks on that account. Well done! 

PR2074

**Review by hetvi11110 — APPROVED** [review]
As mentioned here, I have reviewed this PR, and it is working as expected. Here is a reference video to look at.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13557067/1e063628-037c-46ac-b3d0-34e6f4d68f53

**Comment by one-community** [comment]
Thank you all, moving to final review.

**Review by mohabbasd — DISMISSED** [review]
Hello, I tested you PR and I faced an issue, since I don't have an admin account I tried to do it with a volunteer test account to add the project, task etc. I tried to add the permissions and the permission did not change whether it was owner or volunteer account, and when I try to go back to the dashboard everything goes blank. I tried it multiple times but the same issue keeps happening. Below is a video for your reference.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91215370/5f21a912-86bb-45a7-b2c6-6dfdda8cfad7

**Comment by navneeeth** [comment]
> Hello, I tested you PR and I faced an issue, since I don't have an admin account I tried to do it with a volunteer test account to add the project, task etc. I tried to add the permissions and the permission did not change whether it was owner or volunteer account, and when I try to go back to the dashboard everything goes blank. I tried it multiple times but the same issue keeps happening. Below is a video for your reference.
> 
>  Screen.Recording.2024-03-22.at.2.21.52.AM.mov

Hello, @mohabbasd! Thanks for your review. I notice that in the video you shared, you have granted the wrong permission: Change the Bio Announcement Status. This implementation is for 'Resolve Tasks'. I also observe that your list of permissions in the shared video seem to be outdated. Please use the latest branch on the FE and the latest development branch on the BE to test this.

This implementation is one of many implementations for the list of permissions in the Permissions Management Fixes spreadsheet. It's not meant to be tested on a Volunteer account as a number of the other permissions as listed in the sheet, to be able to view the tasks in the dashboard, and a few others, are yet to be implemented and can lead to troubles testing on a Volunteer account. So, you can test it on an Admin account (by following the steps that I have shared) where the other permissions need not be implemented in order to test this specific implementation.  

Thank you!

**Review by SRamen1999 — COMMENTED** [review]
PR works as expected. I added the permission to admin account and I could see the green tick mark to resolve tasks.

Here I changed the code as specified 
Screen Shot 2024-03-22 at 9 42 43 AM

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/7c8a851e-b5b2-4251-8e31-374efe8b0156

**Comment by rajnada** [comment]
Hii,

I have changed the code as you mentioned in the instructions. And it's working well. Good Job!

Provide permission to Admin Account

image

Here is the task that assigned :

image

image

Thanks

**Review by iven-yao — DISMISSED** [review]
Hi I just review this request.

I login as an admin without resolve task permission, but I can still see green ticks, and able to mark it as done.

following screenshots show that with or without the permission, the task layout is just the same, if I misunderstood this request, please let me know. thank you.
image
image
image
image
image

**Review by yaow62023 — APPROVED** [review]
Hi, i have test your PR fellowing the steps, all the functions work as intended. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/137429365/2543a94a-dc94-4152-92ca-91fb4699359b

**Review by psharma1984 — APPROVED** [review]
hi @navneeeth Clicking the green tick mark the task as resolved!!. Approving

[screen-capture.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123996264/3b1584f2-f8f9-48f3-ae11-de66bbf3300a)

[screen-capture (1).webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123996264/445ed817-6339-41e7-87ca-92089f75a1d7)

**Comment by navneeeth** [comment]
> Hi I just review this request.
> 
> I login as an admin without resolve task permission, but I can still see green ticks, and able to mark it as done.
> 
> following screenshots show that with or without the permission, the task layout is just the same, if I misunderstood this request, please let me know. thank you. image image image image image

Hello, @iven-yao! Yes, admins have this permission available by default--which is why I have provided a code update in the instructions to remove this default case. Please update line 65 as per the instructions and they should not be able to see the green tick (as long as they do not have this permission granted to them). Please try again. Thank you!

**Comment by iven-yao** [comment]
> > Hi I just review this request.
> > I login as an admin without resolve task permission, but I can still see green ticks, and able to mark it as done.
> > following screenshots show that with or without the permission, the task layout is just the same, if I misunderstood this request, please let me know. thank you. image image image image image
> 
> Hello, @iven-yao! Yes, admins have this permission available by default--which is why I have provided a code update in the instructions to remove this default case. Please update line 65 as per the instructions and they should not be able to see the green tick (as long as they do not have this permission granted to them). Please try again. Thank you!

Thank you for pointing out, I tried again and it works perfect as expected.

**Review by iven-yao — APPROVED** [review]
> > Hi I just review this request.
> > I login as an admin without resolve task permission, but I can still see green ticks, and able to mark it as done.
> > following screenshots show that with or without the permission, the task layout is just the same, if I misunderstood this request, please let me know. thank you. image image image image image
> 
> Hello, @iven-yao! Yes, admins have this permission available by default--which is why I have provided a code update in the instructions to remove this default case. Please update line 65 as per the instructions and they should not be able to see the green tick (as long as they do not have this permission granted to them). Please try again. Thank you!

Thank you for pointing out, I tried again and it works perfect as expected.

**Review by Sarthak32 — APPROVED** [review]
Hi I tested this PR as you asked It works well!
Screenshot 2024-03-24 092534
Screenshot 2024-03-24 092546

**Review by Tiantian-C — APPROVED** [review]
The function works as intended. Well done!
Screenshot 2024-03-24 at 7 09 38 PM

**Review by kylepham — APPROVED** [review]
Functionality works as expected. Users with 'Admin' role can resolve their tasks. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/63a06717-98a9-4417-86cb-975c8ef09304

I have a question: does this mean to work only for owner and admin roles? I tried granting a volunteer account the "resolve task" permission but failed to resolve because of unauthorized request. The task agenda doesn't seem to specify this case.

**Comment by Gayathr-2012** [comment]
Hi ,
I reviewed your PR. It works as expected. Good Job.

image

image

image

image

image

image

**Review by SanketKaware — APPROVED** [review]
I have followed the mentioned steps to test this functionality and it works as expected.

After adding the permission to the Admin account, I was able to see the ‘green tick’ icon and resolved the task as well.

Please find my below video reviews for reference

**Before**

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/1b42eb1c-d627-44a5-bed6-b75fc4ccc7ad



**Add Permission**

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/f77f1ec7-f5ff-44b3-92e2-61cec2cc022c


**After**

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/333ab896-53d7-476e-981b-12cc1d3fe7b0

**Pull request merged**